pulse2percept/electrode2currentmap.py

```diff
@@ -570,6 +597,121 @@ def get_pulse(pulse_dur, tsample, interphase_dur, pulsetype):
+    interphasedur : float, optional
+    pulsetype: {'cathodicfirst', 'anodicfirst'}, optional
+
+    """
+
+    # TODO: Default RF size could be half of the electrode gap?
+    def __init__(self, img, implant, coding='amplitude', valrange=[0, 50],
+                 const_amp=20, const_freq=20, rftype='gaussian', rfsize=None,
+                 tsample=0.005 / 1000, dur=0.5, pulsedur=0.5 / 1000.,
+                 interphasedur=0.5 / 1000., pulsetype='cathodicfirst'):
+        try:
+            from skimage.io import imread
+            from skimage.transform import resize
+            from skimage.color import rgb2gray
+        except ImportError:
+            raise ImportError("You do not have scikit-image installed.")
+
+        # Make sure range of values is valid
+        assert len(valrange) == 2 and valrange[1] > valrange[0]
+
+        if isinstance(img, str):
+            # Load image from filename
+            img_large = imread(img, as_grey=True).astype(np.float32)
+        else:
+            # Convert image to grayscale
+            img_large = rgb2gray(np.array(img)).astype(np.float32)
+
+        # Maximize contrast
+        img_large -= img_large.min()
+        img_large /= img_large.max()
+
+        # Center the image on the implant
+        xyr = np.array([[e.x_center, e.y_center, e.radius] for e in implant])
+        xlo = np.min(xyr[:, 0] - xyr[:, 2])
+        xhi = np.max(xyr[:, 0] + xyr[:, 2])
+        ylo = np.min(xyr[:, 1] - xyr[:, 2])
+        yhi = np.max(xyr[:, 1] + xyr[:, 2])
+
+        # Resize the image accordingly and flip up down
+        img_resize = resize(img_large, (yhi - ylo, xhi - xlo))
+        img_resize = np.flipud(img_resize)
+
+        # Make a grid that has the image's coordinates on the retina
+        yg = np.linspace(ylo, yhi, img_resize.shape[0])
+        xg = np.linspace(xlo, xhi, img_resize.shape[1])
+        yg, xg = np.meshgrid(yg, xg)
+
+        # For each electrode, find the stimulation strength (magnitude)
+        magn = []
+        for e in implant:
+            rf = receptive_field(e, xg, yg, rftype, rfsize)
+            magn.append(np.mean(rf * img_resize))
+        magn = np.array(magn)
+
+        # Normalize to valrange
+        magn = (magn - magn.min()) / (magn.max() - magn.min())
+        magn = magn * np.diff(valrange) + valrange[0]
+
+        # Map magnitude to either freq or amp of pulse train
+        pulses = []
+        for m in magn:
+            if coding == 'amplitude':
+                # Map magnitude to amplitude
+                amp = m
+                freq = const_freq
+            elif coding == 'frequency':
+                # Map magnitude to frequency
+                amp = const_amp
+                freq = m
+            else:
+                e_s = "Acceptable values for `coding` are 'amplitude' or"
+                e_s += "'frequency'."
+                raise ValueError(e_s)
+
+            pt = Psycho2Pulsetrain(tsample, freq=freq, amp=amp, dur=dur,
```

Comment:
Maybe there's a way to speed up/save memory by doing these all in the same time-series object? OTOH, might not be worth the hassle.